src/main/scala/com/databricks/labs/overwatch/utils/Structures.scala

```diff
@@ -194,47 +205,64 @@ private[overwatch] class ApiCallFailure(
 
 private[overwatch] class TokenError(s: String) extends Exception(s) {}
 
-private[overwatch] class InvalidInstanceDetailsException(
-                                                          instanceDetails: DataFrame,
-                                                          dfCheck: DataFrame)
+/**
+ * Specialized error to
+ *
+ * @param target          Pipeline Source throwing the error
+ * @param invalidReportDF details of keys causing the error
+ * @param fromCol         column name of the active start date or time
+ * @param untilCol        column name of the expiry date or time, null == active
+ */
+private[overwatch] class InvalidType2Input(
+                                            target: PipelineTable,
+                                            invalidReportDF: DataFrame,
+                                            fromCol: String,
+                                            untilCol: String
+                                          )
   extends Exception() with SparkSessionWrapper {
 
   import spark.implicits._
-  private val erroredRecordsReport = instanceDetails
-    .withColumn("API_Name", lower(trim('API_Name)))
+
+  private val erroredRecordsReport = target.asDF
     .join(
-      dfCheck
-        .select(
-          lower(trim('API_Name)).alias("API_Name"),
-          'rnk, 'rn, 'previousUntil,
-          datediff('activeFrom, 'previousUntil).alias("daysBetweenCurrentAndPrevious")
-        ),
-      Seq("API_Name")
+      invalidReportDF,
+      target.keys
     )
-    .orderBy('API_Name, 'activeFrom, 'activeUntil)
+    .orderBy(target.keys.map(col) ++ Array(col(fromCol), col(untilCol)): _*)
```

Comment:
I see that `target.keys.map(col)` in multiple places. I would just add the function to `PipelineTable` to return seq of columns, something like

```scala
def keyColumns: keys.map(col)
```

Reply:
good idea, let's add that as a feature request.